File: LC/[1778] Shortest Path In A Hidden Grid - recursive.py
from collections import deque


class Solution:
    def findShortestPath(self, master: 'GridMaster') -> int:
        directions = [
            ('U', -1, 0, 'D'),
            ('D', 1, 0, 'U'),
            ('L', 0, -1, 'R'),
            ('R', 0, 1, 'L'),
        ]

        reachable = {(0, 0)}
        target = None

        def dfs(r, c):
            nonlocal target

            if master.isTarget():
                target = (r, c)

            for direction, dr, dc, reverse in directions:
                nr, nc = r + dr, c + dc

                if (nr, nc) in reachable:
                    continue
                if not master.canMove(direction):
                    continue

                reachable.add((nr, nc))

                master.move(direction)
                dfs(nr, nc)
                master.move(reverse)  # Backtrack physically.

        dfs(0, 0)

        if target is None:
            return -1

        queue = deque([(0, 0, 0)])
        visited = {(0, 0)}

        while queue:
            r, c, distance = queue.popleft()

            if (r, c) == target:
                return distance

            for _, dr, dc, _ in directions:
                neighbor = (r + dr, c + dc)

                if neighbor in reachable and neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((*neighbor, distance + 1))

        return -1

File: LC/[1778] Shortest Path In A Hidden Grid - recursive.java
import java.util.*;

class Solution {
    private static final char[] DIR = {'U', 'D', 'L', 'R'};
    private static final char[] BACK = {'D', 'U', 'R', 'L'};
    private static final int[][] DELTA = {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    private final Set<Long> reachable = new HashSet<>();
    private Long target;

    private long key(int row, int col) {
        return ((long) row << 32) | (col & 0xffffffffL);
    }

    public int findShortestPath(GridMaster master) {
        reachable.clear();
        target = null;
        reachable.add(key(0, 0));

        dfs(master, 0, 0);

        if (target == null) {
            return -1;
        }

        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[]{0, 0, 0});
        reachable.remove(key(0, 0));

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0], col = cell[1], distance = cell[2];

            if (key(row, col) == target.longValue()) {
                return distance;
            }

            for (int[] delta : DELTA) {
                int nr = row + delta[0];
                int nc = col + delta[1];

                if (reachable.remove(key(nr, nc))) {
                    queue.offer(new int[]{nr, nc, distance + 1});
                }
            }
        }

        return -1;
    }

    private void dfs(GridMaster master, int row, int col) {
        if (master.isTarget()) {
            target = key(row, col);
        }

        for (int d = 0; d < 4; d++) {
            int nr = row + DELTA[d][0];
            int nc = col + DELTA[d][1];

            if (reachable.contains(key(nr, nc))
                    || !master.canMove(DIR[d])) {
                continue;
            }

            reachable.add(key(nr, nc));
            master.move(DIR[d]);
            dfs(master, nr, nc);
            master.move(BACK[d]); // Restore the robot's position.
        }
    }
}